### DISCUSSION TITLE
Naming convention of our variables and compatibility with Style Dictionary

### DISCUSSION THREAD

**Discussion by owensgit** [Legal-and-General/canopy-design-tokens#1] [General]
The Style Dictionary docs recommend that you structure your tokens by **'Category / Type / Item' - [CTI](https://amzn.github.io/style-dictionary/#/properties?id=category-type-item)**.

For example, take this token for a heading colour:

```
{
  "color": {
    "font": {
      "heading": {
        "value": "#000"
      }
    }
  }
}
```

This generates a token with the name `color-font-heading`, which becomes a CSS var: `--color-font-heading`. In CTI terms, `color` is a **category**, `font` is the **type** and `heading` is the **item**.

The advantage of using this structure is to do with Style Dictionary's built-in **transforms** - these are basically functions that convert your token values (e.g. `#000`) into the required format for each platform. For example, a transform can covert a size/number value into `rem` for CSS, `pt` for iOS and `dp`/`sp` for Android.

Style Dictionary uses the CTI structure to decide how to transform a value - e.g. any token that has a category of `color` will be transformed into the relevant colour formats, a category of `size` will be transformed into sizing/spacing units, etc. 

Our current variable names don't follow this convention, so many of them won't be transformed correctly if we use them as our token names. E.g. `border-width` can't be converted into `rem` or `pt` values out of the box because Style Dictionary doesn't know that it should use the size transform, however `size-border-width` would work.

> Note that the starting value for all sizes is a plain number, and in the case of rems it simply turns the number 1 into `1rem` - but for Android it becomes `16sp`.

This leaves us with a few of options as far as I can see:

**1. Refactor our variable names into CTI structure**
This would, of course, be a huge breaking change! But the advantage being that it's aligned to the base configuration of Style Dictionary and will use many transforms out of the box.

**2. Define our own structure and create our own transform groups**
This will enable us to keep our variable names as close as possible to the original, although some may still have to change - but we'd need to write our own custom transform functions, and transform groups to handle our values (although we should be able to reuse/call some existing ones).

**3. Map CTI based tokens to existing Canopy CSS variables**
This will essentially be a clean slate in the design token codebase to fully follow the CTI structure, but then map the resulting variables to our current CSS variables in the Angular Canopy project. This means no breaking changes, but introduces a maintenance overhead, complexity and eventually inconsistency when other projects use the tokens.

There are still some things that I haven't explored... the Style Dictionary examples repo contains examples of a custom component based structure, which looks a bit like option 2, and there are other things such as base64 encoding of some assets to make them portable between platforms, etc. It's very much first steps in the exploration of how straightforward it is to customise Style Dictionary.

We may also find that a combination of some or all of these options are required. So perhaps a combination of option 1 and 2, where we minimise changing variable names as much as possible, but still need to make some changes in order to avoid a heavily customised setup.

It's super early days, but it would be great to hear some initial thoughts on the above. Is there anything you would probably rule out?

**Comment by andy-polhill** [comment]
Great work looking into this @owensgit.

In style dictionary are you able to reference tokens from other tokens? e.g. something like..

```json
{
  "color": {
    "font": {
      "heading": {
        "value": "--color-super-blue"
      }
    }
  }
}
```
From an initial scan it feels like a balance between a large Breaking Change now that makes our future work easier, or a more custom approach that might be slower in the longer term. My initial instinct is probably to the breaking change, If we were feeling clever we could write a [codemod](https://www.npmjs.com/package/jscodeshift) for it.

**Reply by owensgit** [reply]
Thanks Andy, yes you can reference other tokens, like this:

```json
{
  "color": {
    "font": {
      "heading": {
        "value": "{color.super-blue.value}"
      }
    }
  }
}
```

Thanks for your thoughts, I'm still investigating the advanced examples which contain custom transforms and other customisations that create alternative structures, so it'll be good to see how tricky they can get. My initial instinct is also perhaps to swallow the breaking change for an easier life in the future. I can certainly see the logic in why they split things out into categories like `color`, `size` etc. Btw, I've added a small example of how this could work a [comment below](https://github.com/Legal-and-General/canopy-design-tokens/discussions/1#discussioncomment-506790).

**Comment by owensgit** [comment]
### Example of using default CTI structure

This shows how the current **button** variables could be refactored to follow Style Dictionary's CTI structure - and what CSS variable names they'd result in after build. This might help us visualise the kind of changes we'd need to make it we were to adopt this structure entirely.

**Snippet of the current button variables in Canopy**

```css
:root {
  --btn-disabled-bg-color: var(--color-platinum);
  --btn-disabled-border-color: var(--color-platinum);
  --btn-disabled-color: var(--color-taupe-grey);

  --btn-line-height: 1.5;
  --btn-min-width: 6rem;
  /* etc...  */
}
```

This would be split into the `color` and `size` categories like this...

**Color**

```json
{
  "color": {
    "button": {
      "disabled-bg": { "value": "{color.platinum.value}" },
      "disabled-border": { "value": "{color.platinum.value}" },
      "disabled": { "value": "{color.taupe-grey.value}" }
    }
  }
}
```

**Size**

```json
{
  "size": {
    "button": {
      "line-height": { "value": "1.5" },
      "min-width": { "value": "6" }
    }
  }
}
```

### Result

After build, the CSS variables would end up like this.

```css
:root {
  /* Button colors */
  --color-button-disabled-bg: #e3e3e3;
  --color-button-disabled-border: #e3e3e3;
  --color-button-disabled: #929292;

  /* Button sizes */
  --size-button-line-height: 1.5rem;
  --size-button-min-width: 6rem;
}
```

**Reply by elenagarrone** [reply]
Amazing work @owensgit! I think I would go for the breaking change and use the transformers as they are already setup.

Maybe something we can explore is creating an angular schematic that updates the current css variables to the new ones. So that users can just do `ng update canopy` and it's all done automatically for them.
I would be happy to investigate this option during one of my learning morning.

**Comment by owensgit** [comment]
### Example of defining our own structure and custom transform groups

> Branch: **experiment-custom-transforms** - full example under experiments/ directory

This shows how we could preserve the current Canopy variable names, by using custom [Transform Groups](https://amzn.github.io/style-dictionary/#/transform_groups) in Style Dictionary.

The idea is that you extend the default configuration, creating a custom Transform Function which looks for a particular flag on the design token.

For example, let's look at the Canopy var `--btn-disabled-color`. The design token would look like this:

```
"btn": {
  "disabled": {
    "bg-color": {
      "value": "{color.platinum.value}",
      "group": "color" // flag
    }
  }
}
```

The `group` property is our flag, letting us know we want to transform this value into a colour.

We then create a custom build file; `build.js`, which contains our custom transform functions and transform groups. E.g. this is what the custom colour transform would look like.

```js
/*
 * Register custom transform to handle tokens that have a group
 * attribute of "color".
 */
StyleDictionary.registerTransform({
  name: 'color-group/css',
  type: 'value',
  matcher: function(prop) {
      return prop.group === 'color';
  },
  transformer: function(prop) {
      return StyleDictionary.transform['color/css'].transformer(prop);
  }
});
```

The good thing about the above custom transform is that it simply calls the original `color/css` transform that comes with Style Dictionary, so we're not duplicating the logic, just bypassing the **[CTI](https://amzn.github.io/style-dictionary/#/properties?id=category-type-item)** structure.

Once all the custom parts are added, the rest of the `build.js` file initialises the StyleDictionary methods to extend the default configuration and then run a build. Therefore, instead of using the Style Dictionary CLI command to run a build, you need to `node build.js` and provide a script in package.json.

The upside to this approach is that we don't need to rename our variables.

The downside is that every token will need the `group` flag, rather than be parsed automatically by the default config, and of course it's still a custom layer of config even though it's quite simple.

**Reply by andy-polhill** [reply]
It looks like a good approach, I had a couple of thoughts...
1. Are we doing the simplest possible thing first and shipping that? for example I wonder if we just do the brand colours first and try and ship that end to end. That might get us off the ground and potentially have some value on it's own?
2. The mapping doesn't look too bad, but I think the superior end goal might be to go for the out of the box functionality if it keeps things simple, and if we are ok with the convention it forces. If we think that end goal is better, instead of focusing on the mapping code, we could write a simple [code mod](https://github.com/facebook/codemod) that just helps people with the breaking change? (thought 1 would win in a priority call :-) )

**Reply by owensgit** [reply]
> Are we doing the simplest possible thing first and shipping that? for example I wonder if we just do the brand colours first and try and ship that end to end. That might get us off the ground and potentially have some value on it's own?

I agree, that would also have value for the designers, as we could exporting the brand colours into Sketch palette files.

> The mapping doesn't look too bad, but I think the superior end goal might be to go for the out of the box functionality if it keeps things simple, and if we are ok with the convention it forces.

Yes I think that's probably best option overall too. My only reservation being that with Option 1, variables can feel a little spread out across a few files (e.g. button colours in one place, sizes in another). But the out-of-box config seems better long term.
